orttraining/orttraining/core/framework/ortmodule_graph_builder.cc

```diff
@@ -76,15 +76,17 @@ Status OrtModuleGraphBuilder::Build(const std::vector<std::vector<int64_t>>* inp
     ORT_RETURN_IF_ERROR(SetConcreteInputShapes(*input_shapes_ptr));
   }
 
+  // Optimize the inference graph
+  std::unordered_set<std::string> x_node_arg_names;
+  ORT_RETURN_IF_ERROR(OptimizeForwardGraph(x_node_arg_names));
```

Comment:
We have two graph_transformer_utils.cc files, one is in onnxruntime folder (say, No.1) for any InferenceSession run, one is in orttraining folder (say, No.2) for training only, and it's applied to forward graph only.

For training, it's applying No.2 on forward graph, build gradient graph, and then apply No.1 during InferenceSession running
For inference, it's applying No.1 during InferenceSession running.

We cannot apply No.2 to eval mode because No.2 is for training only, some compute logic is different between training and eval (for example, BatchNorm). If you apply No.2 to eval mode, then during the eval mode, some Ops may run with logic that are for training because of some optimizers, which is not correct. That's the reason we but this OptimizeForwardGraph (which apply No.2 on forward graph) after the if and return. We cannot move it forward.

The LayerNormFusion is part of both No.1 and No.2, so for eval mode, logically the LayerNorm should also be fused when No.1 optimizers are applied. If not, then we need to check why. If it's because the structure of the subgraph is different (between training forward graph and eval graph) so that LayerNormFusion cannot fuse, then we need to check why the subgraph structure is different, or adjust the LayerNormFusion code to cover the new pattern.